src/core/unif.ml

```diff
@@ -36,81 +36,59 @@ let no_problems : problems =
-let instantiation : meta -> term array -> term -> tmbinder Bindlib.box option
-  = fun m ts u ->
+(** [instantiation ctx m ts u] checks whether, in a problem [m[ts]=u], [m] can
+    be instantiated and returns the corresponding instantiation. It does not
+    check whether the instantiation is closed though. *)
+let instantiation : ctxt -> meta -> term array -> term ->
+  tmbinder Bindlib.box option = fun ctx m ts u ->
   if (!can_instantiate || internal m) && not (occurs m u) then
-    match nl_distinct_vars ts with
-    | None -> None
-    | Some vs -> Some (Bindlib.bind_mvar vs (lift u))
+    match nl_distinct_vars ctx ts with
+    | None     -> None
+    | Some(vs) -> Some (Bindlib.bind_mvar vs (lift u))
   else None
 
-(** [instantiate m ts u] check whether, in a problem [m[ts]=u], [m] can be
-   instantiated and, if so, instantiate it. *)
-let instantiate : meta -> term array -> term -> bool = fun m ts u ->
-  match instantiation m ts u with
-  | Some bu when Bindlib.is_closed bu -> set_meta m (Bindlib.unbox bu); true
+(** [instantiate ctx m ts u] check whether, in a problem [m[ts]=u], [m] can be
+    instantiated and, if so, instantiate it. *)
+let instantiate : ctxt -> meta -> term array -> term -> bool =
+  fun ctx m ts u ->
+  match instantiation ctx m ts u with
+  | Some(bu) when Bindlib.is_closed bu -> set_meta m (Bindlib.unbox bu); true
   | _ -> false
 
 (** [solve cfg p] tries to solve the unification problems of [p] and
-   returns the constraints that could not be solved. *)
+    returns the constraints that could not be solved. *)
 let rec solve : problems -> unif_constrs = fun p ->
-  (*if !log_enabled then log_unif "%a" pp_problem p;*)
   match p with
   | { to_solve = []; unsolved = []; _ } -> []
   | { to_solve = []; unsolved = cs; recompute = true } ->
      solve {no_problems with to_solve = cs}
   | { to_solve = []; unsolved = cs; _ } -> cs
-  | { to_solve = (t,u)::to_solve; _ } -> solve_aux t u {p with to_solve}
+  | { to_solve = (c,t,u)::to_solve; _ } -> solve_aux c t u {p with to_solve}
 
 (** [solve_aux t1 t2 p] tries to solve the unificaton problem given by [p] and
     the constraint [(t1,t2)], starting with the latter. *)
-and solve_aux : term -> term -> problems -> unif_constrs = fun t1 t2 p ->
-  let (h1, ts1) = Eval.whnf_stk t1 []
-  and (h2, ts2) = Eval.whnf_stk t2 [] in
-  if !log_enabled then
-    log_unif "solve %a ≡ %a" pp (add_args h1 ts1) pp (add_args h2 ts2);
+and solve_aux : ctxt -> term -> term -> problems -> unif_constrs =
+  fun ctx t1 t2 p ->
+  let t1 = Eval.whnf ctx t1 in
+  let t2 = Eval.whnf ctx t2 in
+  let (h1, ts1) = Basics.get_args t1 in
+  let (h2, ts2) = Basics.get_args t2 in
+  if !log_enabled then log_unif "solve %a" pp_constr (ctx, t1, t2);
 
   let add_to_unsolved () =
-    let t1 = add_args h1 ts1 and t2 = add_args h2 ts2 in
-    if Eval.eq_modulo t1 t2 then solve p
-    else solve {p with unsolved = (t1,t2) :: p.unsolved}
+    if Eval.eq_modulo ctx t1 t2 then solve p else
```

Comment:
From this point, the context `ctx` is completely useless because defined variables of `ctx` have already been unfolded by Eval.whnf in lines 72-73. This means that the ctxt argument of solve_aux could be completely removed by computing the whnf in solve by having in line 66 `| { to_solve = (c,t,u)::to_solve; _ } -> solve_aux (Eval.whnf c t) (Eval.whnf c u) {p with to_solve}`.